server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -893,6 +897,65 @@ public boolean setIgnoredValues(Map<String, List<IgnoredSourceFieldMapper.NameVa
         public String fieldName() {
             return ObjectMapper.this.fullPath();
         }
+
+        interface FieldWriter {
+            void writeTo(XContentBuilder builder) throws IOException;
+
+            record FieldLoader(SourceLoader.SyntheticFieldLoader loader) implements FieldWriter {
+                @Override
+                public void writeTo(XContentBuilder builder) throws IOException {
+                    loader.write(builder);
+                }
+            }
+
+            interface IgnoredSourceWriter extends FieldWriter {
+                FieldWriter mergeWith(IgnoredSourceFieldMapper.NameValue value);
+            }
+
+            record IgnoredSource(IgnoredSourceFieldMapper.NameValue value) implements IgnoredSourceWriter {
+                @Override
+                public void writeTo(XContentBuilder builder) throws IOException {
+                    value.write(builder);
+                }
+
+                @Override
+                public FieldWriter mergeWith(IgnoredSourceFieldMapper.NameValue merged) {
+                    var initialValues = new ArrayList<IgnoredSourceFieldMapper.NameValue>();
+                    initialValues.add(value);
+                    initialValues.add(merged);
+
+                    return new CompositeIgnoredSource(initialValues);
+                }
+            }
+
+            class CompositeIgnoredSource implements IgnoredSourceWriter {
```

Comment:
Nit: I wonder if the interface complexity above is justified, since this is used only within the scope of this class. I'd probably use just this class for now, and move to an interface if needed in the future..